sui-execution/latest/sui-adapter/src/gas_meter.rs

```diff
@@ -0,0 +1,381 @@
+        self.0
+            .charge(1, 1, 2, Type::Bool.size().into(), size_reduction.into())
+    }
+
+    fn charge_vec_pack<'a>(
+        &mut self,
+        _ty: impl TypeView + 'a,
+        args: impl ExactSizeIterator<Item = impl ValueView>,
+    ) -> PartialVMResult<()> {
+        // We will perform `num_args` number of pops.
+        let num_args = args.len() as u64;
+        // The amount of data on the stack stays constant except we have some extra metadata for
+        // the vector to hold the length of the vector.
+        self.0.charge(1, 1, num_args, VEC_SIZE.into(), 0)
+    }
+
+    fn charge_vec_len(&mut self, _ty: impl TypeView) -> PartialVMResult<()> {
+        self.0
+            .charge(1, 1, 1, Type::U64.size().into(), REFERENCE_SIZE.into())
+    }
+
+    fn charge_vec_borrow(
+        &mut self,
+        _is_mut: bool,
+        _ty: impl TypeView,
+        _is_success: bool,
+    ) -> PartialVMResult<()> {
+        self.0.charge(
+            1,
+            1,
+            2,
+            REFERENCE_SIZE.into(),
+            (REFERENCE_SIZE + Type::U64.size()).into(),
+        )
+    }
+
+    fn charge_vec_push_back(
+        &mut self,
+        _ty: impl TypeView,
+        _val: impl ValueView,
+    ) -> PartialVMResult<()> {
+        // The value was already on the stack, so we aren't increasing the number of bytes on the stack.
+        self.0.charge(1, 0, 2, 0, REFERENCE_SIZE.into())
+    }
+
+    fn charge_vec_pop_back(
+        &mut self,
+        _ty: impl TypeView,
+        _val: Option<impl ValueView>,
+    ) -> PartialVMResult<()> {
+        self.0.charge(1, 1, 1, 0, REFERENCE_SIZE.into())
+    }
+
+    fn charge_vec_unpack(
+        &mut self,
+        _ty: impl TypeView,
+        expect_num_elements: NumArgs,
+        _elems: impl ExactSizeIterator<Item = impl ValueView>,
+    ) -> PartialVMResult<()> {
+        // Charge for the pushes
+        let pushes = u64::from(expect_num_elements);
+        // The stack size stays pretty much the same modulo the additional vector size
+        self.0.charge(1, pushes, 1, 0, VEC_SIZE.into())
+    }
+
+    fn charge_vec_swap(&mut self, _ty: impl TypeView) -> PartialVMResult<()> {
+        let size_decrease = REFERENCE_SIZE + Type::U64.size() + Type::U64.size();
+        self.0.charge(1, 1, 1, 0, size_decrease.into())
+    }
+
+    fn charge_drop_frame(
+        &mut self,
+        _locals: impl Iterator<Item = impl ValueView>,
+    ) -> PartialVMResult<()> {
+        Ok(())
+    }
+
+    fn remaining_gas(&self) -> InternalGas {
+        if !self.0.charge {
+            return InternalGas::new(u64::MAX);
+        }
+        self.0.gas_left
+    }
+
+    fn get_profiler_mut(&mut self) -> Option<&mut GasProfiler> {
+        self.0.profiler.as_mut()
+    }
+
+    fn set_profiler(&mut self, profiler: GasProfiler) {
+        self.0.profiler = Some(profiler);
+    }
+}
+
+pub fn abstract_memory_size(status: &GasStatus, val: impl ValueView) -> AbstractMemorySize {
```

Comment:
Note: This needed to be moved out of being a member function on `GasStatus` as it takes a `ValueView` and therefore introduces a dependency on `move-vm-types`